### PR TITLE
feat: apply pep574 out-of-band pickling to DefaultContainer

### PR DESCRIPTION
**Pull request by larme**
## What does this PR address?

This PR is an improvement of #3606 . Instead of using pickle5 to avoid memory copying of a single numpy ndarray, codes improved by this PR can avoid copying of any composite data structure (e.g. a list of ndarray) containing numpy ndarray or pandas dataframe.

Fixes #(issue)

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: tests/unit/_internal/utils/test_pickle.py [hunk 1]
```diff
@@ -23,15 +23,15 @@ def test_pep574_restore() -> None:
     indices: list[int]
     bs, concat_buffer_bs, indices = pep574_dumps(lst)
     restored = t.cast(
-        list["ext.NpNDArray"], pep574_loads(bs, concat_buffer_bs, indices)
+        t.List["ext.NpNDArray"], pep574_loads(bs, concat_buffer_bs, indices)
```
Comment (aarnphm) [inline tests/unit/_internal/utils/test_pickle.py:26]: what is the change here, you can just quote the type `"list['ext.NpNDArray']"` for `t.cast` right?

**Review by aarnphm — DISMISSED** [review]

**Comment by bojiang** [comment]
Potential:
- [ ] pandas DataFrame doing
- [ ] binary file (need investigation, not now)
- [ ] PIL.Image (not supported naturally)

**Comment by bojiang** [comment]
Will continue to review after dataframe supported

**Comment by aarnphm** [comment]
@larme should we also mention here the performance cost for this?

File: src/bentoml/_internal/runner/container.py [hunk 2]
```diff
@@ -487,20 +479,45 @@ def batch_to_batches(
...
+
+        bs: bytes
+        concat_buffer_bs: bytes
+        indices: list[int]
+        bs, concat_buffer_bs, indices = pep574_dumps(batch)
+
+        if indices:
+            meta["with_buffer"] = True
+            data = concat_buffer_bs
+            meta["pickle_bytes_str"] = base64.b64encode(bs).decode("ascii")
```
Comment (bojiang) [inline src/bentoml/_internal/runner/container.py:493]: From larme:
When `with_buffer`=True, the bs will be small. Will ignore the overhead.

**Review by bojiang — APPROVED** [review]

**Pull request merged**